blockchain/chain.go

```diff
@@ -1121,11 +1160,18 @@ func (b *BlockChain) connectBestChain(node *blockNode, block *btcutil.Block, fla
 		// Perform several checks to verify the block can be connected
 		// to the main chain without violating any rules and without
 		// actually connecting the block.
-		view := NewUtxoViewpoint()
-		view.SetBestHash(parentHash)
-		stxos := make([]SpentTxOut, 0, countSpentOutputs(block))
 		if !fastAdd {
-			err := b.checkConnectBlock(node, block, view, &stxos)
+			// We create a viewpoint here to avoid mutating the utxo cache.
+			// The block is not considered valid until checkconnectblock
+			// returns and the mutation would force us to undo the cache.
```

Comment:
I might be misunderstanding the comment or the code, but if this is an invalid block, I think it will still mutate the cache. `checkConnectBlock` calls `fetchInputUtxos` which will call `fetchUtxosFromCache`. Then the block can be invalid on one of the later checks.

Reply:
Pushed a fixup commit to re-word the comment.

I originally made that comment to mean that we can't have new coins added or coins spent until all the validation passes as that would require new code to undo the changes to the utxocache. This is because `checkConnectBlock` adds/spends utxos before the signature validation.

It's true that the utxo cache is being mutated by `fetchUtxosFromCache` but that mutation is ok in that it only loads up utxos from the database. It doesn't spend or add new utxos. This mutation is ok as the worst thing that can happen is extra usage of memory in the cache.